File: IPWorks Samples/WebSocket Server/python/websocket_server.py
# 
# IPWorks 2022 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworks
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworks import *

input = sys.hexversion<0x03000000 and raw_input or input

def ensureArg(args, prompt, index):
  if len(args) <= index:
    while len(args) <= index:
      args.append(None)
    args[index] = input(prompt)
  elif args[index] == None:
    args[index] = input(prompt)

def fireConnected(e):
  global wsserver
  print(wsserver.get_ws_connection_remote_host(e.connection_id) + " has connected.")

def fireDisconnected(e):
  global wsserver
  print("Disconnected from %s" % wsserver.get_ws_connection_remote_host(e.connection_id))

def fireDataIn(e):
  global wsserver
  print("Received a message from %s and the message is: %s" % (wsserver.get_ws_connection_remote_host(e.connection_id), e.text))
  print("Echoing the message back...")
  wsserver.send(e.connection_id, e.text)

print("*****************************************************************\n")
print("* This demo shows how to set up an echo server using WSServer.  *\n")
print("*****************************************************************\n")

port = int(input("Local Port: "))

try:
  wsserver = WSServer()
  wsserver.on_connected = fireConnected
  wsserver.on_disconnected = fireDisconnected
  wsserver.on_data_in = fireDataIn

  buffer = ""
  while buffer == "":
    buffer = input("Use SSL? [y/n]")

  if buffer == "y" or buffer == "Y":
    buffer = ""
    while buffer == "":
      print("Please choose a certificate type:")
      print("1: PFX File")
      print("2: PEMKey File")
      buffer = input("Selection: ")

    if buffer == "1" or buffer == "": #by default use PFX
      wsserver.set_ssl_cert_store_type(2)
    else:
      wsserver.set_ssl_cert_store_type(6)

    buffer = ""
    buffer = input("Enter Certificate File Location (.\sslcert.pfx): ")
    if buffer == "":
      buffer = ".\sslcert.pfx"

    wsserver.set_ssl_cert_store(buffer)

    buffer = ""
    buffer = input("Enter Certificate Password (password): ")
    if buffer == "":
      buffer = "password"

    wsserver.set_ssl_cert_store_password(buffer)
    wsserver.set_ssl_cert_subject("*")
    wsserver.set_use_ssl(True)
  
  wsserver.set_local_port(port)
  wsserver.set_local_host('localhost')
  wsserver.start_listening()
  print("Listening... press Ctrl-C to shutdown")
  while True:
    wsserver.do_events()

except IPWorksError as e:
  print(e)
  print("ERROR: %s" % e.message)

except KeyboardInterrupt:
  print("Shutdown requested...exiting")
  wsserver.shutdown()

File: IPWorks Samples/WebSocket Server/php/websocket_server.php
<?php
/*
 * IPWorks 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/ipworks_wsserver.php');
require_once('../include/ipworks_certmgr.php');
require_once('../include/ipworks_const.php');
?>
<?php
global $wsserver;
class MyWSServer extends IPWorks_WSServer
{
  function FireConnected($param) {
    global $wsserver;
    echo $wsserver->getWSConnectionRemoteHost($param['connectionid']) . " has connected.\n";
  }
  function FireDisconnected($param) {
    echo "Remote host disconnected: " . $param['description'] . "\n";
  }
  function FireDataIn($param) {
    global $wsserver;
    echo "Received '" . $param['text'] . "' from " . $wsserver->getWSConnectionRemoteHost($param['connectionid']) . "\n";
    echo "Echoing message back...\n";
    $wsserver->doSendText($param['connectionid'], $param['text']);
  }
  function FireSSLClientAuthentication($param) {
    $param['accept'] = true;
    return $param;
  }
}

$wsserver = new MyWSServer();
if ($argc < 2) {
  echo "Usage: php websocket_server.php port [certFile certPassword]\n\n";
  echo "  port:          the port on which the server will listen\n"; 
  echo "  certFile:      the certificate file to use for SSL\n";
  echo "  certPassword:  the password for the certificate file\n\n";
  echo "Example: php websocket_client.php 777 test.pfx test\n";
  return;
} else {
  echo "*****************************************************************\n";
  echo "* This demo shows how to set up an echo server using WSServer.  *\n";
  echo "*****************************************************************\n";
  $wsserver->setLocalPort($argv[1]);
  if ($argc == 4) {
    // SSL cert/password is provided.
    $wsserver->setSSLCertStoreType(99); // auto
    $wsserver->setSSLCertStore($argv[2]);
    $wsserver->setSSLCertStorePassword($argv[3]);
    $wsserver->setSSLCertSubject("*");
    $wsserver->setUseSSL(true);
  }
}

function input($prompt) {
  echo $prompt;
  $handle = fopen("php://stdin", "r");
  $data = trim(fgets($handle));
  fclose($handle);
  return $data;
}

try {
  $wsserver->doStartListening();
  echo "Listening on port " . $wsserver->getLocalPort() . "... press Ctrl-C to shutdown.\n";

  while (true) {
    $wsserver->doEvents();
  }
} catch (Exception $ex) {
  echo 'Error: ' . $ex->getMessage() . "\n";
}
?>